src/pp-annotator/modules/PPUI.ts

```diff
@@ -122,100 +122,102 @@ interface IState {
 
 interface IOptions {
   element?: Element;
-  editorExtensions?: Array<{}>;
-  viewerExtensions?: Array<{}>;
 }
 
 /**
  * pp annotator ui module (almost unchanged annotator.ui.main)
  */
 export default class PPUI implements IModule {
   state: IState | undefined;
-  editorExtensions: any;
-  viewerExtensions: any;
   parseRanges: any;
   element: any;
 
   constructor(options?: IOptions) {
     if (typeof options === 'undefined' || options === null) {
       options = {};
     }
-
     this.element = options.element || document.body;
-    this.editorExtensions = options.editorExtensions || [];
-    this.viewerExtensions = options.viewerExtensions || [];
 
     // Local helpers
     this.parseRanges = rangesParser(this.element, '.annotator-hl');
   }
 
   start(app: App) {
+    let state: IState;
     const parseRanges = this.parseRanges;
+    /*
+      State object declarations
+     */
+    const highlighter = new annotator.ui.highlighter.Highlighter(this.element);
+    const textselector = new annotator.ui.textselector.TextSelector(this.element, {
+      onSelection(ranges, event) {
+        if (!state) {
```

Comment:
if you'd write this function as `onSelection: (ranges, event) => { ... },` it should be possible to just use `this.state`, instead of passing a dedicated `state` here.

Unless `this` is required here to be the default bind (however I see no usage of it?)

Reply:
I actually did it this way, but after the modification I thought the readability was much worse afterwards and decided to keep the intermediary state variable. Not sure if you see it this way, too.
It's not a big deal, but I find lines like 
```js
this.state.menu.load(annotation, this.state.interactionPoint);
```
a bit ugly. The reason is there are no other references to `this` in all callbacks like `onSelection` except in `this.state`, and, probably, by design there should be no other references to the `PPUI` instance; only references to other components (which is a flow we should probably change, but then better make a radical change then have a mixture of different approaches) . If so, `this` is  a redundant prefix to `state`.

This is probably more words than it's really worth ;) and well, this is just my feeling about current application flow and this issue remains more of a philosophical issue, as we will change the flow soon anyway.

A full version after the modification, so you can judge yourself:
```js
 const highlighter = new annotator.ui.highlighter.Highlighter(this.element);
    const textselector = new annotator.ui.textselector.TextSelector(this.element, {
      onSelection(ranges, event) {
        if (!this.state) {
          throw new Error('App not initialized!');
        }

        if (!this.state.editor.isShown() && ranges.length > 0) {
          const url = window.location.href;
          const annotation = AnnotationViewModel.fromSelection(parseRanges(ranges), url);
          this.state.interactionPoint = util.mousePosition(event);
          this.state.menu.load(annotation, this.state.interactionPoint);
        } else {
          this.state.menu.hide();
        }
      },
    });
```

Reply:
maybe inside of `onSelection` deconstruct `this.state`, e.g.:
```
const {
  editor,
  menu,
  interactionPoint,
} = this.state;
```
that would remove the issue you've described (bloated `this.state.foo.bar`) still keeping the changes I've suggested 🙂